File: Java/0017-letter-combinations-of-a-phone-number.java
/**
 * Cartesian Product
 * 
 * n: digits length
 *
 * Time: O(n * 4^n)
 *       in the worst-case scenario 4^n combinations are generated since each combination is of length n and characters are copied when constructing the string
 *
 * Space: O(n * 4^n)
 *        4^n result strings each of length n characters are stored 
 *        storing intermediate combinations does not change this growth rate
 * 
 * Backtracking can be used if the goal is to generate the results without storing all intermediate combinations
 */
class Solution {
    private static final Map<Character, List<Character>> DIGIT_TO_LETTERS = Map.of(
        '2', List.of('a', 'b', 'c'),
        '3', List.of('d', 'e', 'f'),
        '4', List.of('g', 'h', 'i'),
        '5', List.of('j', 'k', 'l'),
        '6', List.of('m', 'n', 'o'),
        '7', List.of('p', 'q', 'r', 's'),
        '8', List.of('t', 'u', 'v'),
        '9', List.of('w', 'x', 'y', 'z')
    );

    public List<String> letterCombinations(String digits) {
        List<String> combinations = new ArrayList<>();
        combinations.add("");

        for (char digit : digits.toCharArray()) {
            List<Character> letters = DIGIT_TO_LETTERS.get(digit);
            List<String> nextCombinations = new ArrayList<>();

            for (int i = 0; i < combinations.size(); i++) {
                for (char letter : letters) {
                    nextCombinations.add(combinations.get(i) + letter);
                }
            }

            combinations = nextCombinations;
        }

        return combinations;
    }
}

/**
 * Backtracking
 *
 * n: digits length
 *
 * Time: O(n * 4^n)
 * 
 * Space: O(n * 4^n)
 *        O(n) auxiliary space.
 */
class Solution {
    private static final Map<Character, List<Character>> DIGIT_TO_LETTERS = Map.of(
        '2', List.of('a', 'b', 'c'),
        '3', List.of('d', 'e', 'f'),
        '4', List.of('g', 'h', 'i'),
        '5', List.of('j', 'k', 'l'),
        '6', List.of('m', 'n', 'o'),
        '7', List.of('p', 'q', 'r', 's'),
        '8', List.of('t', 'u', 'v'),
        '9', List.of('w', 'x', 'y', 'z')
    );

    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        StringBuilder sharedState = new StringBuilder();

        backtracking(0, digits, sharedState, result);

        return result;
    }

    private void backtracking(
        int index,
        String digits,
        StringBuilder sharedState,
        List<String> result
    ) {
        // all digits have been processed so the current combination is complete
        if (index == digits.length()) {
            result.add(sharedState.toString());
            return;
        }

        List<Character> letters = DIGIT_TO_LETTERS.get(digits.charAt(index));
        
        for (char letter : letters) {
            sharedState.append(letter);
            backtracking(index + 1, digits, sharedState, result);
            sharedState.deleteCharAt(sharedState.length() - 1);
        }
    }
}

File: Kotlin/0017-letter-combinations-of-a-phone-number.kt
/**
 * Iterative Cartesian Product
 * Time Complexity: O(4^n * n) where n is the length of digits; generates all combinations iteratively.
 * Space Complexity: O(4^n * n) intermediate lists are created at each fold step.
 */
class Solution {
    fun letterCombinations(digits: String): List<String> {
        if (digits.isEmpty()) return emptyList()

        val digitLettersMap = mapOf(
            '2' to "abc",
            '3' to "def",
            '4' to "ghi",
            '5' to "jkl",
            '6' to "mno",
            '7' to "pqrs",
            '8' to "tuv",
            '9' to "wxyz"
        )

        val lettersList = digits.map { digit ->
            digitLettersMap[digit] ?: ""
        }

        return lettersList.fold(listOf("")) { acc, letters ->
            acc.flatMap { prefix ->
                letters.map { char -> prefix + char }
            }
        }
    }
}

/**
 * Backtracking (DFS)
 * Time Complexity: O(4^n * n) explores all possible letter combinations via recursive tree traversal.
 * Space Complexity: O(n) recursion stack depth equals the number of digits
 */
class Solution {
    private val digitMap = arrayOf(
        "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
    )

    fun letterCombinations(digits: String): List<String> {
        if (digits.isEmpty()) return emptyList()

        val result = ArrayList<String>()
        backtrack(digits, 0, StringBuilder(), result)
        return result
    }

    private fun backtrack(
        digits: String,
        index: Int,
        currentStr: StringBuilder,
        result: MutableList<String>
    ) {
        if (index == digits.length) {
            result.add(currentStr.toString())
            return
        }

        val digit = digits[index] - '0' // micro-optimization over .digitToInt()
        val letters = digitMap[digit]

        for (char in letters) {
            currentStr.append(char)
            backtrack(digits, index + 1, currentStr, result)
            currentStr.deleteCharAt(currentStr.length - 1)
        }
    }
}

